File: pong2p/debug_browser.py
#!/usr/bin/env python3
"""
Headless browser debug for pong2p ONNX page.
Run: pip install playwright && playwright install chromium && python debug_browser.py

Loads page, clicks Start, waits for frames, captures console logs and screenshots.
Output: debug-output/

Options: DEBUG_URL=... DEBUG_HEADED=1 DEBUG_WEBGPU=1 python debug_browser.py
"""
import asyncio
import json
import os
import sys
import time
from pathlib import Path

OUT_DIR = Path(__file__).parent / "debug-output"
WAIT_FRAMES = 15
TIMEOUT_MS = 120000
BASE = "https://wendlerc.github.io/pong2p"
HEADED = os.environ.get("DEBUG_HEADED", "").lower() in ("1", "true", "yes")
WEBGPU = os.environ.get("DEBUG_WEBGPU", "").lower() in ("1", "true", "yes")


async def main():
    try:
        from playwright.async_api import async_playwright
    except ImportError:
        print("Install: pip install playwright && playwright install chromium")
        return 1

    OUT_DIR.mkdir(exist_ok=True)
    logs = []

    def log(msg):
        line = f"[{__import__('datetime').datetime.now().isoformat()}] {msg}"
        logs.append(line)
        print(line)

    log(f"Launching Chromium (headed={HEADED})...")
    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=not HEADED)
        context = await browser.new_context()
        page = await context.new_page()

        def on_console(msg):
            log(f"CONSOLE {msg.type}: {msg.text}")

        page.on("console", on_console)

        url = os.environ.get("DEBUG_URL", f"{BASE}/onnx.html" + ("?webgpu=1" if WEBGPU else ""))
        log(f"Navigating to {url}")
        await page.goto(url, wait_until="networkidle", timeout=60000)

        log("Waiting for Start button...")
        await page.wait_for_selector("#startBtn:not([disabled])", timeout=90000)

        log("Clicking Start...")
        await page.click("#startBtn")

        log(f"Waiting ~{WAIT_FRAMES} frames (checking FPS every 2s)...")
        fps_history = []
        start = time.monotonic()
        for i in range(WAIT_FRAMES * 2):
            await asyncio.sleep(2)
            fps_el = await page.query_selector("#fpsValue")
            fps_text = await fps_el.text_content() if fps_el else "?"
            elapsed = (time.monotonic() - start) * 1000
            mem = await page.evaluate("""() => {
                if (performance.memory) return {used: performance.memory.usedJSHeapSize, total: performance.memory.totalJSHeapSize, limit: performance.memory.jsHeapSizeLimit};
                return null;
            }""")
            mem_mb = f", mem={mem['used']/1e6:.1f}MB" if mem else ""
            fps_history.append({"t_ms": int(elapsed), "fps": fps_text, "mem_mb": mem["used"]/1e6 if mem else None})
            log(f"  {elapsed/1000:.1f}s: FPS={fps_text}{mem_mb}")
            if fps_text == "err":
                log("FPS shows err - stopping")
                break
            if elapsed > TIMEOUT_MS:
                break

        log("Taking screenshot...")
        await page.screenshot(path=OUT_DIR / "final.png")

        log("Capturing canvas...")
        canvas_data = await page.evaluate("""() => {
            const c = document.getElementById('frame');
            return c ? c.toDataURL('image/png') : null;
        }""")
        if canvas_data:
            import base64
            b64 = canvas_data.split(",", 1)[-1]
            (OUT_DIR / "canvas.png").write_bytes(base64.b64decode(b64))

        await browser.close()

    (OUT_DIR / "log.txt").write_text("\n".join(logs))
    (OUT_DIR / "fps_history.json").write_text(json.dumps(fps_history, indent=2))
    log(f"Done. Output in {OUT_DIR}")
    return 0


if __name__ == "__main__":
    exit(asyncio.run(main()))

File: pong2p/debug-browser.mjs
#!/usr/bin/env node
/**
 * Headless browser debug for pong2p ONNX page.
 * Run: npx playwright install chromium && node debug-browser.mjs
 * Or: npm install playwright && node debug-browser.mjs
 *
 * Loads page, clicks Start, waits for frames, captures console logs and screenshots.
 * Output: debug-output/
 */
import { chromium } from 'playwright';
import { writeFileSync, mkdirSync, existsSync } from 'fs';
import { join } from 'path';

const BASE = 'https://wendlerc.github.io/pong2p';
const OUT_DIR = join(process.cwd(), 'debug-output');
const WAIT_FRAMES = 15;
const TIMEOUT_MS = 120000;

async function main() {
  mkdirSync(OUT_DIR, { recursive: true });
  const logs = [];
  const log = (msg) => {
    const line = `[${new Date().toISOString()}] ${msg}`;
    logs.push(line);
    console.log(line);
  };

  log('Launching Chromium...');
  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext();
  const page = await context.newPage();

  page.on('console', (msg) => {
    const text = msg.text();
    const type = msg.type();
    log(`CONSOLE ${type}: ${text}`);
  });
  page.on('pageerror', (err) => log(`PAGE ERROR: ${err.message}`));

  const url = process.argv[2] || `${BASE}/onnx.html`;
  log(`Navigating to ${url}`);
  await page.goto(url, { waitUntil: 'networkidle', timeout: 60000 });

  log('Waiting for Start button to be enabled...');
  await page.waitForSelector('#startBtn:not([disabled])', { timeout: 90000 });

  log('Clicking Start...');
  await page.click('#startBtn');

  log(`Waiting for ~${WAIT_FRAMES} frames (checking FPS every 2s)...`);
  const fpsHistory = [];
  const startTime = Date.now();
  for (let i = 0; i < WAIT_FRAMES * 2; i++) {
    await page.waitForTimeout(2000);
    const fpsEl = await page.$('#fpsValue');
    const fpsText = fpsEl ? await fpsEl.textContent() : '?';
    fpsHistory.push({ t: Date.now() - startTime, fps: fpsText });
    log(`  ${(Date.now() - startTime) / 1000}s: FPS=${fpsText}`);
    if (fpsText === 'err') {
      log('FPS shows err - stopping');
      break;
    }
    if (Date.now() - startTime > TIMEOUT_MS) break;
  }

  log('Taking screenshot...');
  await page.screenshot({ path: join(OUT_DIR, 'final.png') });

  log('Capturing canvas as image...');
  const canvasData = await page.evaluate(() => {
    const c = document.getElementById('frame');
    if (!c) return null;
    return c.toDataURL('image/png');
  });
  if (canvasData) {
    const base64 = canvasData.replace(/^data:image\/png;base64,/, '');
    writeFileSync(join(OUT_DIR, 'canvas.png'), Buffer.from(base64, 'base64'));
  }

  await browser.close();

  writeFileSync(join(OUT_DIR, 'log.txt'), logs.join('\n'));
  writeFileSync(join(OUT_DIR, 'fps_history.json'), JSON.stringify(fpsHistory, null, 2));
  log(`Done. Output in ${OUT_DIR}`);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
